### PR TITLE
[air] Fix NoneType error loading TorchCheckpoint through `from_uri`.

### LINKED ISSUE #32284: Torch batch prediction fails
Torch batch prediction fails when reading the checkpoint from uri but will be successful if the checkpoint is from an already trained model in memory.

Stack trace:
```
Traceback (most recent call last):
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/_private/serialization.py", line 369, in deserialize_objects
    obj = self._deserialize_object(data, metadata, object_ref)
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/_private/serialization.py", line 252, in _deserialize_object
    return self._deserialize_msgpack_data(data, metadata_fields)
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/_private/serialization.py", line 207, in _deserialize_msgpack_data
    python_objects = self._deserialize_pickle5_data(pickle5_data)
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/_private/serialization.py", line 197, in _deserialize_pickle5_data
    obj = pickle.loads(in_band)
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/train/torch/torch_checkpoint.py", line 68, in __setstate__
    state["_data_dict"] = self._decode_data_dict(state["_data_dict"])
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/train/torch/torch_checkpoint.py", line 46, in _decode_data_dict
    if ENCODED_DATA_KEY not in data_dict:
TypeError: argument of type 'NoneType' is not iterable
```

### Versions / Dependencies

Ray 2.2.0
Python 3.8.13

### Reproduction script

```
from ray.data.dataset import Dataset
from ray.air.checkpoint import Checkpoint
from ray.train.batch_predictor import BatchPredictor
from ray.train.torch import TorchPredictor
import torch.nn as nn

def create_model():    
    return nn.Sequential(
        nn.Linear(5, 1),
        nn.Sigmoid(),
    )

def predict_me(checkpoint: Checkpoint) -> Dataset:
    batch_predictor = BatchPredictor.from_checkpoint(
        checkpoint=checkpoint,
        predictor_cls=TorchPredictor,model=create_model()
    )
    return batch_predictor.predict(data=test_dataset)

uri = "some_uri"
result = TorchTrainer( ... ).fit()
result.checkpoint.to_uri(uri)

checkpoint1 = result.checkpoint
predict_me(checkpoint1) # runs successfully

checkpoint2 = TorchCheckpoint.from_uri(uri)
predict_me(checkpoint2) # fails
```

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by xwjiang2010**
- encode/decode logic of TorchCheckpoint is only relevant in the context of training
- it's currently intercepting `__getstate__` and `__setstate__`
- training worker first sends TorchCheckpoint to Trainable (through ser/deser). The TorchCheckpoint that Trainable receives is already "decoded"
- the last statement may not hold true anymore if training worker uploads TorchCheckpoint directly with `to_uri`, at which point we should revisit where to put the encode/decode logic.

Signed-off-by: xwjiang2010 <xwjiang2010@gmail.com>

## Related issue number
Closes #32284

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by Yard1 — CHANGES_REQUESTED** [review]
I don't think we need this, the model should be already encoded when it was put into the URI. I think the only thing that needs to be fixed is the missing check for the case where `_data_dict` is None:

```python
    def __setstate__(self, state: dict):
        if state.get("_data_dict", None):
            state = state.copy()
            state["_data_dict"] = self._decode_data_dict(state["_data_dict"])
        super().__setstate__(state)
```

**Review by krfricke — APPROVED** [review]
This looks good!

**Comment by xwjiang2010** [comment]
> I don't think we need this, the model should be already encoded when it was put into the URI. I think the only thing that needs to be fixed is the missing check for the case where `_data_dict` is None:
> 
> ```python
>     def __setstate__(self, state: dict):
>         if state.get("_data_dict", None):
>             state = state.copy()
>             state["_data_dict"] = self._decode_data_dict(state["_data_dict"])
>         super().__setstate__(state)
> ```

isn't this for *decoding* ? We encode it when putting into the URI and decode it when retrieving?

**Comment by xwjiang2010** [comment]
waiting for https://github.com/ray-project/ray/pull/32479/files to be landed first...

File: python/ray/train/torch/torch_checkpoint.py [hunk 1]
```diff
@@ -78,7 +78,8 @@ def __getstate__(self) -> dict:
     def __setstate__(self, state: dict):
         if "_data_dict" in state:
             state = state.copy()
-            state["_data_dict"] = self._decode_data_dict(state["_data_dict"])
+            if state["_data_dict"]:
```
Comment (Yard1) [inline python/ray/train/torch/torch_checkpoint.py:81]: I'd move this if to the outer if to avoid an unnecessary shallow copy

Comment (xwjiang2010) [reply python/ray/train/torch/torch_checkpoint.py]: ah right, thanks for catching this!

**Review by Yard1 — APPROVED** [review]

File: python/ray/train/tests/test_torch_predictor.py [hunk 2]
```diff
@@ -16,8 +16,8 @@
 from ray.train.batch_predictor import BatchPredictor
 from ray.train.predictor import TYPE_TO_ENUM
 from ray.train.torch import TorchCheckpoint, TorchPredictor
-
 from ray.train.tests.dummy_preprocessor import DummyPreprocessor
+from ray.tune.tests.conftest import *  # noqa
```
Comment (Yard1) [inline python/ray/train/tests/test_torch_predictor.py:20]: I don't think you need to import from conftest, it should work automatically

**Pull request merged**